spring-cloud-stream-binders/spring-cloud-stream-binder-gemfire/src/main/java/org/springframework/cloud/stream/binder/gemfire/MessageKey.java

```diff
@@ -0,0 +1,229 @@
+/*
+ * Copyright 2015 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.stream.binder.gemfire;
+
+import java.io.ByteArrayOutputStream;
+import java.io.Externalizable;
+import java.io.IOException;
+import java.io.ObjectInput;
+import java.io.ObjectOutput;
+import java.io.ObjectOutputStream;
+
+import com.gemstone.gemfire.cache.EntryOperation;
+import com.gemstone.gemfire.cache.PartitionResolver;
+
+import org.springframework.messaging.Message;
+
+/**
+ * Key class for storing {@link Message messages} in a GemFire region.
+ * This class consists of the following fields:
+ * <ul>
+ *     <li>A unique sequence id generated by the client; this id
+ *         is used to determine the order of messages</li>
+ *     <li>A producer id which is a combination of the producer PID
+ *         and the timestamp of when the producer process was started.</li>
+ *     <li>An integer used as the hash code for GemFire to select a
+ *         bucket for storing the message in a partitioned region.</li>
+ * </ul>
+ *
+ * @author Patrick Peralta
+ */
+public class MessageKey implements Externalizable, Comparable<MessageKey>,
+		PartitionResolver<MessageKey, Message<?>> {
+
+	/**
+	 * Unique sequential id for message.
+	 */
+	private long sequenceId;
+
+	/**
+	 * Encoded identifier for producer. First (leftmost) 16 bits are
+	 * reserved for the PID of the producer process, rightmost 48
+	 * bits are reserved for the timestamp of when the producer
+	 * process was started.
+	 */
+	private long producerId;
+
+	/**
+	 * Object returned for {@link #getRoutingObject}.
+	 */
+	private int routingHash;
+
+	/**
+	 * Constructor for serialization.
+	 */
+	public MessageKey() {
```

Comment:
yep, that's what I feared as soon as I understood we were talking native java serialization. Too bad ;)